### DISCUSSION TITLE
gameRepository와 추가적으로 필요한 in-memory 저장 정보들

### DISCUSSION THREAD

**Discussion by khk811** [42MVP/back-end#56] [Ideas]
roomId를 key, 어떤 게임 정보가 담긴 객체가 value인 gameRepository가 있다고 가정을 한 후, 게임 코드를 1차적으로 작성중입니다.
그러면서 front와 back이 어떻게 동작할지 로직을 어느정도 구상을 했습니다.

- front는 렌더만 합니다. 유저가 키 입력을 한다면 키 입력 소켓 이벤트를 back에 보냅니다.
- back은 게임 정보가 담긴 객체를 특정 timeInterval마다 (eg. 10ms) 각각의 게임 방에 보냅니다. front의 키 입력 이벤트를 감지해서 객체 내 좌표값을 업데이트 합니다.

1. 이때, back은 존재하는 game 방을 순회하면서 모든 방에, 각각의 소켓 연결 상태가 어떻든 상관없이 계속 메시지를 보내기로 했습니다.
이렇다보니 gameRepository에서 어떤 특정 키값으로 게임 객체를 찾는것 뿐만 아니라 gameRepository가 가진 전체 데이터가 필요해졌습니다.

2. 그리고 키 입력 이벤트를 받을때 소켓 아이디만 parameter로 받는게 좋을것 같다는 의견이 있었습니다.
방 아이디 변조 위험도 있을 수 있고, 기본적으로 아이디에 대한 데이터는 서버에서 관리를 하는 편이 좋다고 생각했기 때문입니다.
그래서 소켓 아이디만을 가지고 키 입력 이벤트를 처리하고자 하고,  그러다 보니 <소켓 아이디, 방 아이디>로 데이터 쌍을 가지는 Map이 필요해졌습니다.
(`소켓 아이디로 방을 찾고 -> 그 방으로 게임 객체를 찾고 -> 그 객체 내 paddle 좌표를 업데이트 한다` 이런 흐름으로 구상을 했습니다.)

---

### 요약
1. gameRepository를 만드신다면 map 그 자체를 반환하는 함수를 작성해주실수 있나요?
2. game.gateway.ts 내부 혹은 repository로 Map<socketId, roomId>인 게임 방 Map(혹은 레포지토리)를 만들어도 될까요?

제가 구상한 로직 및 사용하려고 하는 데이터가 불필요하다고 느끼시거나 좀 더 좋은 방법이 생각나신다면 가감없이 말씀해주세요. 수정하겠습니다 🙏

**Comment by sinryuji** [comment]
1. ok
2. 애초에 그냥 게임 레포지토리 자체가 Map<socketId, gameInfo Obejct> 구조로 가는건 어떨까요? gameInfo Object에 내부 필드로 roomId를 들고 있구요

**Comment by khk811** [comment]
### 디스코드 회의 후 결론
1.은 게임 시작시 setInterval()을 하고 렌더 정보를 callback 함수로 호출하는 방식으로 접근하는걸로 변경할 시 불필요합니다.
2. 형키님이 말씀해 주신것 처럼 레포지토리를 Map<socketId, gameInfo Object> 구조로 구현을 해주시면 매우 감사합니다.
